File: src/commonMain/kotlin/me/kolterdyx/compiler/expression/VariableExpression.kt
package me.kolterdyx.compiler.expression

import me.kolterdyx.compiler.Token
import me.kolterdyx.compiler.ValueType
import me.kolterdyx.compiler.ast.ExpressionVisitor

class VariableExpression(
    val name: Token,
    valueType: ValueType
) : Expression(valueType) {
    override fun <R> accept(visitor: ExpressionVisitor<R>): R {
        return visitor.visitVariable(this)
    }

}

File: internal/ast/variableExpression.go
package ast

import (
	"github.com/Kolterdyx/mcbasic/internal/interfaces"
	"github.com/Kolterdyx/mcbasic/internal/tokens"
)

type VariableExpr struct {
	Name tokens.Token
	ResolvedType
}

func (v *VariableExpr) Accept(visitor ExpressionVisitor) any {
	return visitor.VisitVariable(v)
}

func (v *VariableExpr) Type() NodeType {
	return VariableExpression
}

func (v *VariableExpr) ToString() string {
	return v.Name.Lexeme
}

func (v *VariableExpr) GetSourceLocation() interfaces.SourceLocation {
	return v.Name.SourceLocation
}
